### PR TITLE
Add recycler stats

### PR DESCRIPTION
**Pull request by sakridge**

### CODE REVIEW

**Comment by garious** [comment]
Do you guys grok the issues surrounding implementing Drop using Weak refs? Seem like the fundamental problem here is that we never did that piece, and this PR further adds complexity to help manager the existing situation instead of eliminating the possibility of this class of bugs with Weak refs.

**Comment by rob-solana** [comment]
have we assured ourselves we only have 2 recyclers in the system? one for packets, one for blobs?  the stages sometimes ship blobs one way...

**Comment by rob-solana** [comment]
I don't understand Drop using Weak refs, so I don't understand how such would help.

File: src/packet.rs [hunk 1]
```diff
@@ -194,13 +205,29 @@ impl<T: Default> Clone for Recycler<T> {
...
 }
 
+fn try_inc(x: &RefCell<u64>) {
+    if let Ok(mut x) = x.try_borrow_mut() {
+        *x += 1;
+    }
+}
+
 impl<T: Default + Reset> Recycler<T> {
+    pub fn set_name(&mut self, name: String) {
```
Comment (rob-solana) [inline src/packet.rs]: name can be a 'static &str...

**Comment by garious** [comment]
So `skipped_count` might be leaked count?  If `Packet<T>` holds a `Weak` to `PacketRecyler` and then `Arc<RwLock<Packet<T>>>` implements `Drop`, its `drop()` could pass the packet to `recycle()`. Then you'd never need to pass recyclers around and `skip_count()` should drop to zero.

**Comment by garious** [comment]
Implementing `Weak` still feels like a research project. Don't let this PR be blocked by that. If this is what makes the most sense in current code, let's keep moving forward.

File: src/packet.rs [hunk 2]
```diff
@@ -180,12 +180,22 @@ pub enum BlobError {
...
     #[cfg_attr(feature = "cargo-clippy", allow(type_complexity))]
     gc: Arc<Mutex<Vec<(Arc<RwLock<T>>, &'static str)>>>,
+    allocated_count: Arc<AtomicUsize>,
+    recycled_count: Arc<AtomicUsize>,
+    reuse_count: Arc<AtomicUsize>,
+    skipped_count: Arc<AtomicUsize>,
+    name: String,
 }
 
 impl<T: Default> Default for Recycler<T> {
```
Comment (garious) [inline src/packet.rs]: This is what I'd imagine the default implementation would look like. Did you try adding `#[derive(Clone)]` to the struct?

Comment (sakridge) [reply src/packet.rs]: Just tried..

```
error[E0599]: no method named `clone` found for type `packet::Recycler<packet::Packets>` in the current scope
   --> src/rpu.rs:57:29
    |
57  |             packet_recycler.clone(),
    |                             ^^^^^
    |
   ::: src/packet.rs:225:1
    |
225 | pub struct Recycler<T> {
    | ---------------------- method `clone` not found for this
    |
    = note: the method `clone` exists but the following trait bounds were not satisfied:
            `packet::Recycler<packet::Packets> : std::clone::Clone`
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `clone`, perhaps you need to implement it:
            candidate #1: `std::clone::Clone`
```

Comment (sakridge) [reply src/packet.rs]: oh. maybe i need `use std::clone::Clone` in those files...

Comment (sakridge) [reply src/packet.rs]: hmmm. no that wasn't it.

**Pull request merged**